File: examples/php-laravel/tests/Unit/ReferenceConfigTest.php
<?php

namespace Tests\Unit;

use App\ReferenceConfig;
use InvalidArgumentException;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;

final class ReferenceConfigTest extends TestCase
{
    /** @return array<string, string> */
    private function values(): array
    {
        return [
            'listen_host' => '127.0.0.1', 'port' => '15500', 'public_origin' => 'http://127.0.0.1:15500', 'gateway_url' => 'http://127.0.0.1:15501',
            'redis_url' => 'redis://127.0.0.1:6379', 'session_secret' => str_repeat('x', 32), 'session_lifetime_seconds' => '1200',
            'instance_name' => 'php-laravel-a', 'topology' => 'non-ha', 'redis_instance_prefix' => 'cormier:test',
            'redis_session_key_prefix' => 'sessions', 'allowed_tenants' => 'tenant-a', 'allowed_users' => 'user-a',
            'shared_asset_root' => __DIR__, 'sdk_asset_root' => __DIR__,
        ];
    }

    public function test_it_parses_typed_configuration(): void
    {
        $config = ReferenceConfig::fromArray($this->values());
        self::assertSame('127.0.0.1', $config->listenHost);
        self::assertSame(15500, $config->port);
        self::assertTrue($config->allows('tenant-a', 'user-a'));
        self::assertSame('cormier:test:sessions:id', $config->sessionKey('id'));
    }

    public function test_it_rejects_unsafe_origins(): void
    {
        $values = $this->values();
        $values['public_origin'] = 'file:///tmp';
        $this->expectException(InvalidArgumentException::class);
        ReferenceConfig::fromArray($values);
    }

    public function test_it_rejects_explicit_default_origin_ports(): void
    {
        $values = $this->values();
        $values['public_origin'] = 'https://example.test:443';
        $this->expectException(InvalidArgumentException::class);
        ReferenceConfig::fromArray($values);
    }

    public function test_it_rejects_noncanonical_origin_casing(): void
    {
        $values = $this->values();
        $values['public_origin'] = 'https://EXAMPLE.TEST';
        $this->expectException(InvalidArgumentException::class);
        ReferenceConfig::fromArray($values);
    }

    #[DataProvider('forbiddenOriginProvider')]
    public function test_it_rejects_each_forbidden_origin_component(string $origin): void
    {
        $values = $this->values();
        $values['public_origin'] = $origin;
        $this->expectException(InvalidArgumentException::class);
        ReferenceConfig::fromArray($values);
    }

    /** @return array<string, array{string}> */
    public static function forbiddenOriginProvider(): array
    {
        return [
            'credentials' => ['https://user:password@example.test'],
            'query' => ['https://example.test?route=unsafe'],
            'fragment' => ['https://example.test#unsafe'],
        ];
    }

    public function test_it_consumes_canonical_contracts(): void
    {
        $schema = file_get_contents(__DIR__.'/../../../shared-web/reference-app.schema.json');
        $sdk = file_get_contents(__DIR__.'/../../../../sdk/typescript/dist/version.json');
        $protocol = file_get_contents(__DIR__.'/../../../../protocol/fixtures/v1/envelopes.json');
        self::assertIsString($schema);
        foreach (['LISTEN_HOST', 'PORT', 'PUBLIC_ORIGIN', 'GATEWAY_URL', 'REDIS_URL', 'SESSION_SECRET', 'PUBLIC_ROOT'] as $key) {
            self::assertStringContainsString('"'.$key.'"', $schema);
        }
        self::assertStringContainsString('"protocolVersion": "1.0"', (string) $sdk);
        self::assertStringContainsString('"protocolVersion": "1.0"', (string) $protocol);
    }
}

File: examples/ruby-rails/test/reference_config_test.rb
require "json"
require "minitest/autorun"
require "stringio"
require_relative "../lib/reference_config"
require_relative "../lib/limited_body_reader"

class ReferenceConfigTest < Minitest::Test
  def values
    {
      "LISTEN_HOST" => "127.0.0.1",
      "PORT" => "15500",
      "PUBLIC_ORIGIN" => "http://127.0.0.1:15500",
      "GATEWAY_URL" => "http://127.0.0.1:15501",
      "REDIS_URL" => "redis://127.0.0.1:6379",
      "SESSION_SECRET" => "x" * 32,
      "SESSION_LIFETIME_SECONDS" => "1200",
      "INSTANCE_NAME" => "ruby-rails-a",
      "TOPOLOGY" => "non-ha",
      "REDIS_INSTANCE_PREFIX" => "cormier:test",
      "REDIS_SESSION_KEY_PREFIX" => "sessions",
      "ALLOWED_TENANTS" => "tenant-a",
      "ALLOWED_USERS" => "user-a"
    }
  end

  def test_typed_configuration
    config = ReferenceConfig.load(values)
    assert_equal "127.0.0.1", config.listen_host
    assert_equal 15_500, config.port
    assert config.allows?("tenant-a", "user-a")
    assert_equal "cormier:test:sessions:id", config.session_key("id")
    root = File.expand_path("../../..", __dir__)
    assert_equal File.join(root, "examples/shared-web/wwwroot"), config.shared_asset_root
    assert_equal File.join(root, "sdk/typescript/dist"), config.sdk_asset_root
  end

  def test_rejects_unsafe_origin
    environment = values.merge("PUBLIC_ORIGIN" => "file:///tmp")
    assert_raises(ArgumentError) { ReferenceConfig.load(environment) }
  end

  def test_rejects_explicit_default_origin_port
    environment = values.merge("PUBLIC_ORIGIN" => "https://example.test:443")
    assert_raises(ArgumentError) { ReferenceConfig.load(environment) }
  end

  def test_rejects_noncanonical_origin_casing
    environment = values.merge("PUBLIC_ORIGIN" => "https://EXAMPLE.TEST")
    assert_raises(ArgumentError) { ReferenceConfig.load(environment) }
  end

  def test_canonical_contracts
    root = File.expand_path("../../..", __dir__)
    schema = JSON.parse(File.read(File.join(root, "examples/shared-web/reference-app.schema.json")))
    sdk = JSON.parse(File.read(File.join(root, "sdk/typescript/dist/version.json")))
    protocol = JSON.parse(File.read(File.join(root, "protocol/fixtures/v1/envelopes.json")))
    assert_includes schema.fetch("required"), "PUBLIC_ORIGIN"
    assert_equal "1.0", sdk.fetch("protocolVersion")
    assert_equal "1.0", protocol.fetch("protocolVersion")
  end

  def test_limited_body_reader_accepts_the_boundary
    body = "a" * (64 * 1_024)
    assert_equal body, LimitedBodyReader.read(StringIO.new(body), limit: body.bytesize)
  end

  def test_limited_body_reader_rejects_a_chunked_oversized_body
    stream = StringIO.new("a" * (64 * 1_024) + "b")
    assert_raises(LimitedBodyReader::TooLarge) do
      LimitedBodyReader.read(stream, limit: 64 * 1_024)
    end
  end
end
